### PR TITLE
Bump engines and some dependencies

### PR DESCRIPTION
**Pull request by bakkot**
This is a breaking change, but node 16 is no longer supported, so whatever. We claimed to support node 8 but haven't been testing on it in forever.

This isn't quite the latest typescript because typescript-eslint hasn't updated to support it yet.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -72,22 +72,22 @@
...
+		"typescript": "^5.5.4"
 	},
 	"prettier": {
 		"singleQuote": true,
 		"arrowParens": "avoid",
 		"printWidth": 100
 	},
 	"engines": {
-		"node": ">= 12 || ^11.10.1 || ^10.13 || ^8.10"
+		"node": ">= 18"
```
Comment (ljharb) [inline package.json:91]: even if nothing else changed, this alone requires a semver-major bump, fwiw

Comment (bakkot) [reply package.json:91]: Yeah, that's why I said "This is a breaking change".

File: tsconfig.test.json [hunk 2]
```diff
@@ -2,6 +2,7 @@
     "extends": "./tsconfig",
     "include": ["lib/**/*"],
     "compilerOptions": {
+        "outDir": "this needs to be set so it does not ignore lib",
```
Comment (michaelficarra) [inline tsconfig.test.json:5]: wow

**Review by michaelficarra — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (ljharb) [reply package.json:91]: gotcha, the "so whatever" implied you weren't going to publish it as one.

Comment (bakkot) [reply package.json:91]: Aha, I see how that was misleading; I just meant that it's unlikely to break anyone in practice.

**Pull request merged**